File: src/main/java/com/back/mozu/domain/queue/service/LockService.java
package com.back.mozu.domain.queue.service;

import com.back.mozu.global.redis.RedisUtil;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

@Slf4j
@Service
@RequiredArgsConstructor
public class LockService {

    private final RedisUtil redisUtil;

    // 락의 최대 유지 시간: 3초
    // 최대 유지 시간 내에 작업 처리 미완료 시 데드락 방지를 위해 자동으로 락이 해제
    private static final int MAX_LOCK_TIME = 3000;

    // 분산 락 설정
    public boolean acquireLock(String timeSlotId, String token) {
        String lockKey = RedisUtil.lockKey(timeSlotId);
        boolean isAcquired = redisUtil.lockAcquire(lockKey, token, MAX_LOCK_TIME);

        if (isAcquired) {
            log.info("[Lock 획득 성공] timeSlotId: {}, token: {}", timeSlotId, token);
        } else {
            log.warn("[Lock 획득 실패 - 이미 점유됨] timeSlotId: {}, token: {}", timeSlotId, token);
        }

        return isAcquired;
    }

    // 분산 락 해제
    public void releaseLock(String timeSlotId, String token) {
        String lockKey = RedisUtil.lockKey(timeSlotId);

        // RedisUtil 내부의 Lua 스크립트를 통해 내 토큰일 경우 삭제
        redisUtil.lockRelease(lockKey, token);

        log.info("[Lock 해제 완료] timeSlotId: {}, token: {}", timeSlotId, token);
    }
}

File: src/main/java/com/back/mozu/domain/queue/service/LockService.kt
package com.back.mozu.domain.queue.service

import org.slf4j.LoggerFactory
import org.springframework.data.redis.core.StringRedisTemplate
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import java.time.Duration
import java.util.concurrent.ConcurrentHashMap

private const val LOCK_TTL_SECONDS = 10L
private const val WATCHDOG_INTERVAL_MS = 3_000L
private const val MAX_WATCHDOG_EXTENSIONS = 5

@Service
class LockService(
    private val redisTemplate: StringRedisTemplate,
) {
    private val activeLocks = ConcurrentHashMap<String, Pair<String, Int>>()

    fun acquireLock(lockKey: String, lockToken: String): Boolean {
        val acquired = redisTemplate.opsForValue()
            .setIfAbsent(lockKey, lockToken, Duration.ofSeconds(LOCK_TTL_SECONDS))
            ?: false

        if (acquired) {
            activeLocks[lockKey] = Pair(lockToken, 0)
        }
        return acquired
    }

    fun releaseLock(lockKey: String, lockToken: String) {
        val currentToken = redisTemplate.opsForValue().get(lockKey)
        if (currentToken == lockToken) {
            redisTemplate.delete(lockKey)
        }
        activeLocks.remove(lockKey)
    }

    @Scheduled(fixedDelay = WATCHDOG_INTERVAL_MS)
    fun renewActiveLocks() {
        activeLocks.forEach { (lockKey, tokenAndCount) ->
            val (lockToken, extensionCount) = tokenAndCount

            if (extensionCount >= MAX_WATCHDOG_EXTENSIONS) {
                log.warn(
                    "Watchdog 최대 연장 횟수 초과. Lock 강제 만료 예정: lockKey={}, extensionCount={}",
                    lockKey, extensionCount,
                )
                activeLocks.remove(lockKey)
                return@forEach
            }

            val currentToken = redisTemplate.opsForValue().get(lockKey)
            if (currentToken == lockToken) {
                redisTemplate.expire(lockKey, Duration.ofSeconds(LOCK_TTL_SECONDS))
                activeLocks[lockKey] = Pair(lockToken, extensionCount + 1)
                log.debug(
                    "Lock TTL 연장: lockKey={}, extensionCount={}/{}",
                    lockKey, extensionCount + 1, MAX_WATCHDOG_EXTENSIONS,
                )
            } else {
                activeLocks.remove(lockKey)
            }
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger(LockService::class.java)
    }
}